datasets/ua/war_sanctions/crawler.py

```diff
@@ -0,0 +1,348 @@
+            dob = dob_pob[0]
+            h.apply_date(person, "birthDate", dob)
+    if positions:
+        for position in h.multi_split(positions, [" / "]):
+            person.add("position", position)
+    person.add("topics", "poi")
+
+    sanction = h.make_sanction(context, person)
+    sanction.add("reason", " ".join(data.pop("Reasons")))
+    sanction.add("sourceUrl", link)
+
+    context.emit(person, target=True)
+    context.emit(sanction)
+    context.audit_data(data, ignore=["Sanction Jurisdictions"])
+
+
+def crawl_company(context: Context, link):
+    detail_page = context.fetch_html(link, cache_days=3)
+    details_container = detail_page.find(".//main")
+    data = {}
+    for row in details_container.findall(".//div[@class='row']"):
+        divs = row.findall("div")
+        if len(divs) == 2:
+            label_elem, value_elem = divs
+            if "yellow" in label_elem.get("class"):
+                label, value = extract_label_value_pair(label_elem, value_elem, data)
+                data[label] = value
+    name = data.pop("Name")
+    name_abbr = data.pop("Abbreviated name of the legal entity", None)
+    reg_num = data.pop("Registration number")
+
+    company = context.make("Company")
+    company.id = context.make_id(name, name_abbr, reg_num)
+    company.add("name", name)
+    company.add("name", name_abbr)
+    company.add("registrationNumber", reg_num)
+    company.add("address", data.pop("Address"))
+    company.add("country", data.pop("Country"))
+    company.add("taxNumber", data.pop("Tax Number"))
+    company.add("sourceUrl", data.pop("Links"))
+    archive_links = data.pop("Archive links", None)
+    if archive_links is not None:
+        company.add("sourceUrl", archive_links)
+
+    company.add("topics", "poi")
+    sanction = h.make_sanction(context, company)
+    sanction.add("reason", " ".join(data.pop("Reasons")))
+    sanction.add("sourceUrl", link)
+
+    context.emit(company, target=True)
+    context.emit(sanction)
+    context.audit_data(data, ignore=["Sanction Jurisdictions"])
+
+
+def extract_next_page_url(doc):
+    # next page <a> element extraction using xpath
+    next_link_element = doc.xpath("//ul[@class='pagination']//li[@class='next']/a")
+    if next_link_element:
+        next_link = next_link_element[0]
+        return next_link.get("href")
+
+    return None
+
+
+def crawl(context: Context):
+    main_page = context.fetch_html(context.data_url, cache_days=3)
+    node = main_page.find(
+        ".//section[@class='sections d-flex flex-wrap align-items-stretch justify-content-center mb-5 pl-5 pr-5 medium']"
+    )
+    h.assert_dom_hash(node, "dbb9a924d940b3a69a132a102e04dcf0f9fbfc5e")
+
+    for link_info in LINKS:
+        base_url = link_info["url"]
+        data_type = link_info["type"]
+        current_url = base_url
+        visited_pages = 0
+        while current_url:
+            doc = context.fetch_html(current_url, cache_days=3)
```

Comment:
my experience is that it's best not to cache index pages because additions/removals can result in skipped entities